modules/pubsub/broker.go

```diff
@@ -0,0 +1,105 @@
+package pubsub
+
+import (
+	"sync"
+	"errors"
+
+	"github.com/spaceuptech/space-cloud/model"
+	"github.com/spaceuptech/space-cloud/utils"
+)
+
+// type unsubscribe func()(error)
+
+type pubsubBroker interface {
+	Publish(subject string, msg *model.PubsubMsg) error
+	QueueSubscribe(subject, queue string, ch chan *model.PubsubMsg) (func()(error), error)
+}
+
+type pubsubSubscription struct {
+	unsubscribeFunc func()(error)
+	ch              chan *model.PubsubMsg
+}
+
+func (m *Module) Publish(subject string, data interface{}) error {
+	subject = utils.SingleLeadingTrailing(subject, "/")
+	return m.connection.Publish(subject, &model.PubsubMsg{subject, data})
+}
+
+func (m *Module) Subscribe(clientId, subject string, cb model.PubsubCallback) (error) {
+	return m.QueueSubscribe(clientId, subject, clientId, cb)
+}
+
+func (m *Module) QueueSubscribe(clientId, subject, queue string, cb model.PubsubCallback) (error) {
+	receiveChannel := make(chan *model.PubsubMsg)
+	subject = utils.SingleLeadingTrailing(subject, "/")
+	subs, err := m.connection.QueueSubscribe(subject, queue, receiveChannel)
+	if err != nil {
+		return err
+	}
+	subscription := pubsubSubscription{subs, receiveChannel}
+	err = m.storeSubs(subject, clientId, &subscription)
+	if err != nil {
+		subscription.unsubscribeFunc()
+		close(subscription.ch)
+		return err
+	}
+	go func() {
+		for msg := range receiveChannel {
+        	cb(msg)
+		}
+	}()
+	return nil
+}
+
+func (m *Module) Unsubscribe(clientId, subject string) error {
+	subject = utils.SingleLeadingTrailing(subject, "/")
+	subjM, ok := m.storage.Load(subject)
+	if ok {
+		subjectMap := subjM.(sync.Map)
+		subs, ok := subjectMap.Load(clientId)
+		if ok {
+			subscription := subs.(pubsubSubscription)
```

Comment:
You are strong this object as a pointer but retrieving it the other way

Reply:
Oh. Right